core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -223,8 +223,14 @@ class ReplicaManager(val config: KafkaConfig,
       case Some(partition) =>
         if(deletePartition) {
           val removedPartition = allPartitions.remove((topic, partitionId))
-          if (removedPartition != null)
+          if (removedPartition != null) {
             removedPartition.delete() // this will delete the local log
+            val topicPartitionCount = allPartitions.keys.count {
+              case (t, p) => topic == t
```

Comment:
I think the following is a bit better (we stop as soon as we find one partition for `topic` instead of always going through the whole collection.

```scala
val topicHasPartitions = allPartitions.keys.exists { case (t, _) => topic == t }
if (!topicHasPartitions) BrokerTopicStats.removeMetrics(topic)
```